File: pset2/crack/crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

int main(int argc, string argv[])
{
    if(argc != 2)
    {
        printf("Usage: ./crack hashofPassword\n");
        return -1;
    }

    string password = argv[1];

    string allCharacter = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz\0";
    char tmp[6], tmpSalt[3];

    int all = strlen(allCharacter);

    tmpSalt[0] = password[0];
    tmpSalt[1] = password[1];
    tmpSalt[2] = '\0';

    //printf("%s\n", crypt("ROFL", "50"));
    for(int i = 0; i < all - 1; i++)
    {
        tmp[0] = allCharacter[i];
        for(int j = 0; j < all; j++)
        {
            tmp[1] = allCharacter[j];
            for(int k = 0; k < all; k++)
            {
                tmp[2] = allCharacter[k];
                for(int l = 0; l < all; l++)
                {
                    tmp[3] = allCharacter[l];
                    for(int m = 0; m < all; m++)
                    {
                        tmp[4] = allCharacter[m];
                        if(strcmp(crypt(tmp, tmpSalt), password) == 0)
                        {
                            printf("%s\n", tmp);
                            return 0;
                        }
                    }
                }
            }
        }
    }
}

File: pset6/crack/crack.py
import crypt
import sys
import string
from itertools import permutations


def main():
    # Ensure that there are enough command-line arguments
    if(len(sys.argv) != 2):
        print("Usage: python crack.py hashofPassword")

    password = sys.argv[1]

    # Store all alphabets
    allCharacter = string.ascii_letters
    salt = password[:2]

    length = len(allCharacter)
    tmp = [''] * 4

    # Get all permutations of 1, 2, 3, 4, and 5 length
    # strings and store them
    a = permutations(allCharacter, 1)
    b = permutations(allCharacter, 2)
    c = permutations(allCharacter, 3)
    d = permutations(allCharacter, 4)
    e = permutations(allCharacter, 5)

    # Try all the permutations with crypt
    for word in a:
        if(crypt.crypt(''.join(word), salt) == password):
            print(''.join(word))
            exit(0)

    # Try all the permutations with crypt
    for word in b:
        if(crypt.crypt(''.join(word), salt) == password):
            print(''.join(word))
            exit(0)

    # Try all the permutations with crypt
    for word in c:
        if(crypt.crypt(''.join(word), salt) == password):
            print(''.join(word))
            exit(0)

    # Try all the permutations with crypt
    for word in d:
        if(crypt.crypt(''.join(word), salt) == password):
            print(''.join(word))
            exit(0)

    # Try all the permutations with crypt
    for word in e:
        if(crypt.crypt(''.join(word), salt) == password):
            print(''.join(word))
            exit(0)


if __name__ == "__main__":
    main()